Add Statistics options hotkeys

rob-v opened the pull request:
Closes #13743
New hotkeys for Statistics dialog:

![image](https://user-images.githubusercontent.com/16348750/29002908-52bd988c-7aad-11e7-84f8-5c2f2676dfad.png)

pchote commented:
Adding a dependency on #13706 and #13711 which rework the way hotkeys are defined and used.

rob-v commented:
Updated to use new hotkeys design.

pchote commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/ObserverStatsLogic.cs:
```diff
@@ -325,5 +332,44 @@ class StatsDropDownOption
+			if (e.Event == KeyInputEvent.Down && !e.IsRepeat)
+			{
+				var h = Hotkey.FromKeyInput(e);
+				if (h == Game.Settings.Keys.StatisticsBasicKey)
```
These need to be pulled from the chrome logic args (see https://github.com/OpenRA/OpenRA/pull/13781/commits/12c5d05728eeaa405b95f3940cf2de1704904a2d#diff-9c03e754ef92a76206eb1b2c35d8dfaeR68) because these hardcoded `Game.Settings.Keys.*` fields are all going away.

rob-v replied:
sorry, I pushed only rebased stuff, forgot to stage before commit. I don't use Keys.StatisticsBasicKey,... anymore.

rob-v commented:
Updated again, pushed only rebased stuff before. Now pushed also with adaptations for new hotkey design.

pchote commented:
Note for reviewing: https://github.com/OpenRA/OpenRA/pull/13781/commits/029f37680d7ce460097293324ef868679bb62f1c?w=1 is the meat of this PR, ignoring the whitespace and my prereq changes.

The code changes look correct, but I haven't yet reviewed the behavior or choice of keys.

Commit:
Add Statistics options hotkeys

netnazgul commented:
@rob-v can this be prepped for Next+1?

pchote commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/ObserverStatsLogic.cs:
```diff
@@ -325,5 +357,44 @@ class StatsDropDownOption
+			if (e.Event == KeyInputEvent.Down && !e.IsRepeat)
+			{
+				var h = Hotkey.FromKeyInput(e);
+				if (h == statisticsBasicKey.GetValue())
```
If you put these keys in an array then you could reduce duplication by looping over them.  Compare with #13966.

rob-v commented:
Updated with hotkeys array as suggested.

pchote commented:
Sorry, but this needs a rebase to avoid another #14028 situation.

pchote commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/ObserverStatsLogic.cs:
```diff
@@ -146,6 +118,9 @@ public ObserverStatsLogic(World world, WorldRenderer worldRenderer, Widget widge
 				};
+
+			var keyhandler = statsDropDown.Get<LogicKeyListenerWidget>("STATS_DROPDOWN_KEYHANDLER");
+			keyhandler.OnKeyPress = HandleKeyPress;
```
This needs to change to `keyhandler.AddHandler(HandleKeyPress);`

pchote commented on OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs:
```diff
@@ -501,7 +501,12 @@ Action InitInputPanel(Widget panel)
 					{ "ReplaySpeedFastKey", "Fast speed" },
-					{ "ReplaySpeedMaxKey", "Maximum speed" }
+					{ "ReplaySpeedMaxKey", "Maximum speed" },
+					{ "StatisticsBasicKey", "Statistics Basic" },
```
Can we please change these to "Basic Statistics tab" and so on, so they read as proper english?

pchote reviewed (commented):
Looks good overall, just two minor points (of which one is my fault) and a scope creep: 

This feature would be much more useful if the tab hotkeys worked directly from the world without opening the panel first.  It shouldn't be too hard to adapt:
https://github.com/OpenRA/OpenRA/blob/d99a47245622eb627235c510dce42a3d6bfd07d3/OpenRA.Mods.Common/Widgets/Logic/Ingame/MenuButtonsChromeLogic.cs#L84-L89

to open from any of the keys by following the approach of:

https://github.com/OpenRA/OpenRA/blob/d99a47245622eb627235c510dce42a3d6bfd07d3/OpenRA.Mods.Common/Widgets/Logic/Ingame/CommandBarLogic.cs#L205-L220

rob-v commented:
Updated.

pchote commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/MenuButtonsChromeLogic.cs:
```diff
@@ -85,8 +85,29 @@ public MenuButtonsChromeLogic(Widget widget, World world)
+					{
+						var ks = Game.Settings.Keys;
+						var key = new Hotkey(e.Key, e.Modifiers);
+						Hotkey[] statsKeys = { ks.StatisticsBasicKey, ks.StatisticsEconomyKey, ks.StatisticsProductionKey, ks.StatisticsCombatKey, ks.StatisticsGraphKey };
```
Can you please use NamedHotkey instead? These hardcoded ks.* fields are about to move to yaml.

rob-v replied:
I am not sure how to do it.
1. move Statistics<name>Key from ObserverStatsLogic to MenuButtonsChromeLogic and then pass them to ObserverStatsLogic 
2. use ObserverStatsLogic.StatsHotkeys in MenuButtonsChromeLogic. Trying to find now how to get reference to this Logic object in MenuButtonsChromeLogic
3. ?

pchote replied:
Hmm, good question.  I think the best approach here will be to define/accept the same Statistics*Key args in `MenuButtonsChromeLogic` that you use in `ObserverStatsLogic` -- define them in both places rather than reference or try to pass them between classes.

rob-v commented:
Updated.

pchote commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/MenuButtonsChromeLogic.cs:
```diff
@@ -85,8 +94,28 @@ public MenuButtonsChromeLogic(Widget widget, World world)
 			}
+
+			var keyListener = widget.GetOrNull<LogicKeyListenerWidget>("OBSERVER_KEY_LISTENER");
+			if (keyListener != null)
```
Style nit: if a single statement includes its own inner blocks, then the outer block should have braces to avoid ambiguity.

pchote commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/ObserverStatsLogic.cs:
```diff
@@ -14,37 +14,49 @@
+		readonly IEnumerable<Player> players;
+		readonly World world;
+		readonly WorldRenderer worldRenderer;
+		readonly List<StatsDropDownOption> statsDropDownOptions;
```
This can be an array instead of a list.

pchote reviewed (commented):
Looks good, thanks.  Just two more minor nits, and then can you also change the `OBSERVER_STATS_BUTTON` `Key:`'s to use `StatisticsBasic` instead of hardcoding `F1`?

rob-v commented:
Updated. All good ideas. Now much more useful to use hotkeys ingame to open directly a stats tab, though it is little annoying (discouraging to use) that it takes second(s) (since latest playtests) to open a/Stats dialog.

Re OBSERVER_STATS_BUTTON's Key - I removed it, as StatisticsBasicKey already works without this now. Is it ok or you prefer to keep it?

pchote commented:
> though it is little annoying (discouraging to use) that it takes second(s) (since latest playtests) to open a/Stats dialog.

What do you mean?  Is this a recent regression?

> Re OBSERVER_STATS_BUTTON's Key - I removed it, as StatisticsBasicKey already works without this now. Is it ok or you prefer to keep it?

Keep it please so that it will show the hotkey in the tooltip.

rob-v commented:
Updated.
I assumed it is well known, but didn't see more comments about this... #13526
Right, tooltip.

pchote commented on OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs:
```diff
@@ -502,7 +502,12 @@ Action InitInputPanel(Widget panel)
+					{ "StatisticsEconomyKey", "Economy Statistics" },
+					{ "StatisticsProductionKey", "Production Statistics" },
+					{ "StatisticsCombatKey", "Combat Statistics" },
+					{ "StatisticsGraphKey", "Graph Statistics" }
```
"Graph Statistics" doesn't make sense.  Can you please change this to "Economy Graph" or "Statistics Graph"?

pchote commented on OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs:
```diff
@@ -502,7 +502,12 @@ Action InitInputPanel(Widget panel)
 					{ "ReplaySpeedFastKey", "Fast speed" },
-					{ "ReplaySpeedMaxKey", "Maximum speed" }
+					{ "ReplaySpeedMaxKey", "Maximum speed" },
+					{ "StatisticsBasicKey", "Basic Statistics" },
```
Please use lower case "s" for consistency with (most of) the other definitions.
Don't worry about the other ones (e.g. "All Players"), I'll fix those in the pr that pulls these out to the yaml.

pchote commented on mods/ts/chrome/ingame-observer.yaml:
```diff
@@ -18,7 +25,6 @@ Container@OBSERVER_WIDGETS:
 			Height: 25
 			Text: Statistics (F1)
```
Can you please remove the "(F1)" from here too?

pchote reviewed (commented):
Sorry, a couple more and hopefully final set of polish issues:

rob-v commented:
Updated.

reaperrr commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/b5db0048bc8a49426dc39bae8209705b4d4d7778...d8db0ddb0ba698381017eb69ce789edbe94cdc55

netnazgul commented:
Hm, so this was merged 3 weeks ago but didn't get into release?

pchote commented:
This was merged to bleed, not prep.

netnazgul commented:
yes, @abcdefg30 already explained it in discord. So it will be re-checked for the future playtests, right?

abcdefg30 commented:
This PR, the same as all other PRs merged into bleed until a new prep branch is split off, go into the next playtest, yes.
